server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java

```diff
@@ -165,14 +166,21 @@ public <T> T retryTransaction(final TransactionCallback<T> callback, final int q
 
   public final boolean isTransientException(Throwable e)
   {
-    return e != null && (e instanceof RetryTransactionException
-                         || e instanceof SQLTransientException
-                         || e instanceof SQLRecoverableException
-                         || e instanceof UnableToObtainConnectionException
-                         || e instanceof UnableToExecuteStatementException
-                         || connectorIsTransientException(e)
-                         || (e instanceof SQLException && isTransientException(e.getCause()))
-                         || (e instanceof DBIException && isTransientException(e.getCause())));
+    if (e == null) {
+      return false;
+    }
+    if (e.getMessage() != null && e.getMessage().contains(MAX_ALLOWED_PACKET_ERROR)) {
```

Comment:
Is there a more robust way to detect this? Perhaps a particular exception type, or `SQLState` value, or `vendorCode`?

Reply:
For MySQL, the SQLState seems to be `S1000`, error code is `0`, and the root cause exception is `PacketTooBigException`.

```
org.skife.jdbi.v2.exceptions.CallbackFailedException: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: com.mysql.jdbc.PacketTooBigException: Packet for query is too large (9952 > 1024). You can change this value on the server by setting the max_allowed_packet' variable.
```

The solution for this case is simply to check if the cause of `UnableToExecuteStatementException`, which we should always do anyway because `UnableToExecuteStatementException` is not transient by itself. Since `PacketTooBigException` is not transient by any of our other rules, it would be categorized correctly.

Reply:
Unfortunately, this is not the exception we had seen in prod. I tried reproducing it by keeping a high `max_allowed_packet` on the mysql server but a low value in the connection url, but that that again resulted in the same `PacketTooLargeException`. In prod, we had seen a nesting of `CallbackFailedException` -> `UnableToExecuteStatementException` -> `SQLTransientConnectionException`.

Reply:
> Always look at the cause of UnableToExecuteStatementException to decide if it is transient. (this takes care of the mysql driver case)

Sounds good to me.

> For SQLTransientConnectionException, check the exception chain. If any of the causes is a SQLNonTransientException exception, the overall exception is also non-transient.

I'm a little worried about collateral damage from the strategy of looking for `SQLNonTransientException` under ` SQLTransientConnectionException`. How about a more targeted change where we look for `org.mariadb.jdbc.internal.util.exceptions.MaxAllowedPacketException` specifically? I'd be OK with that.

Reply:
Done.
Added a new method `connectorIsNonTransientException` to allow connector implementations to classify an exception as definitely non-transient.